File: src/rlm_research_planner/services/paid_value.py
from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from math import ceil
from typing import Iterable

from rlm_research_planner.domain.models import PaidItem, PaidOffer, PaidValuation


SPEEDUP_ITEM_KINDS = (
    "general",
    "research",
    "training",
    "construction",
    "healing",
    "merging",
    "crafting",
)
PAID_GOALS = (
    "all_round",
    "account_growth",
    "research",
    "construction",
    "troop_training",
    "combat",
    "monster_hunt",
    "equipment",
    "familiar",
    "artifact",
    "heroes",
    "events",
    "resources",
)
PAID_ITEM_KINDS = SPEEDUP_ITEM_KINDS + (
    "gems",
    "monster_common",
    "monster_uncommon",
    "monster_rare",
    "monster_epic",
    "monster_legendary",
    "chest",
    "resource",
    "material",
    "combat_item",
    "boost_item",
    "building_material",
    "familiar_item",
    "monster_energy",
    "hero_item",
    "artifact_item",
    "event_item",
    "currency",
    "custom",
)
DEFAULT_POINTS_EACH = {
    "monster_common": 1.0,
    "monster_uncommon": 4.0,
    "monster_rare": 16.0,
    "monster_epic": 64.0,
    "monster_legendary": 256.0,
}
SPEEDUP_GEM_VALUE_BY_SECONDS = {
    60: 5.0,
    15 * 60: 70.0,
    60 * 60: 130.0,
    3 * 60 * 60: 300.0,
    8 * 60 * 60: 650.0,
    15 * 60 * 60: 1_000.0,
    24 * 60 * 60: 1_500.0,
    3 * 24 * 60 * 60: 4_400.0,
    7 * 24 * 60 * 60: 10_000.0,
    30 * 24 * 60 * 60: 40_000.0,
}
MERGING_SPEEDUP_GEM_VALUE_BY_SECONDS = {
    15 * 60: 140.0,
    60 * 60: 260.0,
    3 * 60 * 60: 600.0,
    8 * 60 * 60: 1_300.0,
    15 * 60 * 60: 2_000.0,
    24 * 60 * 60: 3_000.0,
    3 * 24 * 60 * 60: 8_800.0,
    7 * 24 * 60 * 60: 20_000.0,
}


@dataclass(frozen=True)
class GemSpeedupPurchase:
    gems: int
    purchased_seconds: int


@lru_cache(maxsize=1)
def _standard_speedup_gem_costs() -> tuple[tuple[int, ...], tuple[int, ...]]:
    """Build the cheapest Gem Mall allocation for up to one 30-day item."""

    presets = tuple(
        sorted(
            (
                max(1, int(duration_seconds) // 60),
                max(0, int(gem_cost)),
            )
            for duration_seconds, gem_cost in SPEEDUP_GEM_VALUE_BY_SECONDS.items()
            if duration_seconds > 0 and gem_cost > 0
        )
    )
    maximum_minutes = max(duration for duration, _cost in presets)
    costs = [0] + [10**18] * maximum_minutes
    purchased_minutes = [0] * (maximum_minutes + 1)
    for target_minutes in range(1, maximum_minutes + 1):
        best_cost = 10**18
        best_minutes = 0
        for duration, cost in presets:
            previous = max(0, target_minutes - duration)
            candidate_cost = cost + costs[previous]
            candidate_minutes = duration + purchased_minutes[previous]
            if (candidate_cost, candidate_minutes) < (best_cost, best_minutes):
                best_cost = candidate_cost
                best_minutes = candidate_minutes
        costs[target_minutes] = best_cost
        purchased_minutes[target_minutes] = best_minutes
    return tuple(costs), tuple(purchased_minutes)


def minimum_gems_for_speedup_seconds(seconds: int) -> GemSpeedupPurchase:
    """Return the cheapest standard Gem Mall speed-ups covering ``seconds``."""

    required_minutes = max(0, ceil(max(0, int(seconds)) / 60))
    if required_minutes <= 0:
        return GemSpeedupPurchase(0, 0)
    costs, purchased_minutes = _standard_speedup_gem_costs()
    maximum_minutes = len(costs) - 1
    maximum_cost = int(SPEEDUP_GEM_VALUE_BY_SECONDS[maximum_minutes * 60])
    full_items, remainder = divmod(required_minutes, maximum_minutes)
    if remainder <= 0:
        return GemSpeedupPurchase(
            full_items * maximum_cost,
            full_items * maximum_minutes * 60,
        )
    with_remainder = GemSpeedupPurchase(
        full_items * maximum_cost + costs[remainder],
        (full_items * maximum_minutes + purchased_minutes[remainder]) * 60,
    )
    with_extra_full_item = GemSpeedupPurchase(
        (full_items + 1) * maximum_cost,
        (full_items + 1) * maximum_minutes * 60,
    )
    return min(
        (with_remainder, with_extra_full_item),
        key=lambda value: (value.gems, value.purchased_seconds),
    )


@dataclass(frozen=True)
class PaidOfferSummary:
    speedup_seconds: dict[str, int]
    total_speedup_seconds: int
    included_gems: int
    item_gem_value: float
    speedup_gem_value: float
    total_gem_value: float
    direct_item_points: float
    speedup_points: float
    total_points: float
    points_per_diamond: float | None
    gems_per_diamond: float | None


def paid_kind_has_time(kind: str) -> bool:
    return kind in SPEEDUP_ITEM_KINDS


def default_points_each(kind: str) -> float:
    return float(DEFAULT_POINTS_EACH.get(kind, 0.0))


def default_gem_value_each(kind: str) -> float:
    return 1.0 if kind == "gems" else 0.0


def speedup_gem_value_each(kind: str, duration_seconds: int) -> float:
    duration = max(0, int(duration_seconds))
    if kind == "merging":
        return MERGING_SPEEDUP_GEM_VALUE_BY_SECONDS.get(duration, 0.0)
    if kind == "crafting":
        return 0.0
    return SPEEDUP_GEM_VALUE_BY_SECONDS.get(duration, 0.0)


def _speedup_rate(valuation: PaidValuation, kind: str) -> float:
    return max(
        0.0,
        float(
            {
                "general": valuation.general_speedup_points_per_hour,
                "research": valuation.research_speedup_points_per_hour,
                "training": valuation.training_speedup_points_per_hour,
                "construction": valuation.construction_speedup_points_per_hour,
                "healing": valuation.healing_speedup_points_per_hour,
                "merging": valuation.merging_speedup_points_per_hour,
                "crafting": valuation.crafting_speedup_points_per_hour,
            }.get(kind, 0.0)
        ),
    )


def summarize_paid_offer(
    offer: PaidOffer,
    valuation: PaidValuation,
) -> PaidOfferSummary:
    speedup_seconds = {kind: 0 for kind in SPEEDUP_ITEM_KINDS}
    item_gem_value = 0.0
    speedup_gem_value = 0.0
    direct_item_points = 0.0
    for item in offer.items:
        quantity = max(0, int(item.quantity))
        if item.kind in speedup_seconds:
            speedup_seconds[item.kind] += max(0, int(item.duration_seconds)) * quantity
            if valuation.use_speedup_gem_presets and item.gem_value_each <= 0:
                speedup_gem_value += (
                    speedup_gem_value_each(item.kind, item.duration_seconds)
                    * quantity
                )
        item_gem_value += max(0.0, float(item.gem_value_each)) * quantity
        direct_item_points += max(0.0, float(item.points_each)) * quantity
    included_gems = max(0, int(offer.included_gems)) + max(
        0, int(offer.bonus_gems)
    )
    total_gem_value = included_gems + item_gem_value + speedup_gem_value
    speedup_points = sum(
        seconds / 3600.0 * _speedup_rate(valuation, kind)
        for kind, seconds in speedup_seconds.items()
    )
    total_points = (
        total_gem_value * max(0.0, float(valuation.points_per_gem))
        + direct_item_points
        + speedup_points
    )
    cost = max(0, int(offer.diamond_cost))
    return PaidOfferSummary(
        speedup_seconds=speedup_seconds,
        total_speedup_seconds=sum(speedup_seconds.values()),
        included_gems=included_gems,
        item_gem_value=item_gem_value,
        speedup_gem_value=speedup_gem_value,
        total_gem_value=total_gem_value,
        direct_item_points=direct_item_points,
        speedup_points=speedup_points,
        total_points=total_points,
        points_per_diamond=total_points / cost if cost else None,
        gems_per_diamond=total_gem_value / cost if cost else None,
    )


def sorted_paid_offers(
    offers: Iterable[PaidOffer],
    valuation: PaidValuation,
) -> tuple[PaidOffer, ...]:
    return tuple(
        sorted(
            offers,
            key=lambda offer: (
                -(
                    summarize_paid_offer(offer, valuation).points_per_diamond
                    or 0.0
                ),
                -summarize_paid_offer(offer, valuation).total_points,
                offer.title.casefold(),
            ),
        )
    )

File: tools/RLMResearchPlannerPWA/src/paid-value.js
export const SPEEDUP_ITEM_KINDS = [
  "general", "research", "training", "construction", "healing", "merging", "crafting",
];
export const PAID_GOALS = ["all_round", "account_growth", "research", "construction", "troop_training", "combat", "monster_hunt", "equipment", "familiar", "artifact", "heroes", "events", "resources"];
export const PAID_ITEM_KINDS = [
  ...SPEEDUP_ITEM_KINDS,
  "gems",
  "monster_common",
  "monster_uncommon",
  "monster_rare",
  "monster_epic",
  "monster_legendary",
  "chest",
  "resource",
  "material",
  "combat_item",
  "boost_item",
  "building_material",
  "familiar_item",
  "monster_energy",
  "hero_item",
  "artifact_item",
  "event_item",
  "currency",
  "custom",
];

const DEFAULT_POINTS_EACH = {
  monster_common: 1,
  monster_uncommon: 4,
  monster_rare: 16,
  monster_epic: 64,
  monster_legendary: 256,
};
const SPEEDUP_GEM_VALUE_BY_SECONDS = new Map([
  [60, 5], [15 * 60, 70], [60 * 60, 130], [3 * 60 * 60, 300],
  [8 * 60 * 60, 650], [15 * 60 * 60, 1000], [24 * 60 * 60, 1500],
  [3 * 24 * 60 * 60, 4400], [7 * 24 * 60 * 60, 10000], [30 * 24 * 60 * 60, 40000],
]);
let standardSpeedupGemCosts = null;
const MERGING_SPEEDUP_GEM_VALUE_BY_SECONDS = new Map([
  [15 * 60, 140], [60 * 60, 260], [3 * 60 * 60, 600], [8 * 60 * 60, 1300],
  [15 * 60 * 60, 2000], [24 * 60 * 60, 3000], [3 * 24 * 60 * 60, 8800],
  [7 * 24 * 60 * 60, 20000],
]);
export const PAID_OFFER_DOCUMENT_TYPE = "RLMResearchPlanner.paid-offers";

const finite = (value, fallback = 0) => {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
};
const nonNegative = (value, fallback = 0) => Math.max(0, finite(value, fallback));

function buildStandardSpeedupGemCosts() {
  if (standardSpeedupGemCosts) return standardSpeedupGemCosts;
  const presets = [...SPEEDUP_GEM_VALUE_BY_SECONDS.entries()]
    .map(([seconds, gems]) => [Math.max(1, Math.trunc(seconds / 60)), Math.max(0, Math.trunc(gems))])
    .sort((left, right) => left[0] - right[0]);
  const maximumMinutes = presets.at(-1)[0];
  const costs = Array(maximumMinutes + 1).fill(Number.MAX_SAFE_INTEGER);
  const purchasedMinutes = Array(maximumMinutes + 1).fill(0);
  costs[0] = 0;
  for (let targetMinutes = 1; targetMinutes <= maximumMinutes; targetMinutes += 1) {
    for (const [duration, cost] of presets) {
      const previous = Math.max(0, targetMinutes - duration);
      const candidateCost = cost + costs[previous];
      const candidateMinutes = duration + purchasedMinutes[previous];
      if (candidateCost < costs[targetMinutes]
        || (candidateCost === costs[targetMinutes] && candidateMinutes < purchasedMinutes[targetMinutes])) {
        costs[targetMinutes] = candidateCost;
        purchasedMinutes[targetMinutes] = candidateMinutes;
      }
    }
  }
  standardSpeedupGemCosts = { costs, purchasedMinutes, maximumMinutes };
  return standardSpeedupGemCosts;
}

export function minimumGemsForSpeedupSeconds(seconds) {
  const requiredMinutes = Math.max(0, Math.ceil(nonNegative(seconds) / 60));
  if (!requiredMinutes) return { gems: 0, purchasedSeconds: 0 };
  const { costs, purchasedMinutes, maximumMinutes } = buildStandardSpeedupGemCosts();
  const maximumCost = SPEEDUP_GEM_VALUE_BY_SECONDS.get(maximumMinutes * 60);
  const fullItems = Math.floor(requiredMinutes / maximumMinutes);
  const remainder = requiredMinutes % maximumMinutes;
  if (!remainder) {
    return {
      gems: fullItems * maximumCost,
      purchasedSeconds: fullItems * maximumMinutes * 60,
    };
  }
  const withRemainder = {
    gems: fullItems * maximumCost + costs[remainder],
    purchasedSeconds: (fullItems * maximumMinutes + purchasedMinutes[remainder]) * 60,
  };
  const withExtraFullItem = {
    gems: (fullItems + 1) * maximumCost,
    purchasedSeconds: (fullItems + 1) * maximumMinutes * 60,
  };
  return withRemainder.gems < withExtraFullItem.gems
    || (withRemainder.gems === withExtraFullItem.gems
      && withRemainder.purchasedSeconds <= withExtraFullItem.purchasedSeconds)
    ? withRemainder
    : withExtraFullItem;
}

export function defaultPaidValuation() {
  return {
    pointsPerGem: 1,
    generalSpeedupPointsPerHour: 0,
    researchSpeedupPointsPerHour: 0,
    trainingSpeedupPointsPerHour: 0,
    constructionSpeedupPointsPerHour: 0,
    healingSpeedupPointsPerHour: 0,
    mergingSpeedupPointsPerHour: 0,
    craftingSpeedupPointsPerHour: 0,
    useSpeedupGemPresets: true,
  };
}

export function emptyPaidOffer() {
  const now = new Date().toISOString();
  return {
    offerId: "",
    title: "",
    goal: "all_round",
    memo: "",
    diamondCost: 0,
    includedGems: 0,
    bonusGems: 0,
    items: [],
    createdAt: now,
    updatedAt: now,
  };
}

export function paidKindHasTime(kind) {
  return SPEEDUP_ITEM_KINDS.includes(kind);
}

export function defaultPointsEach(kind) {
  return DEFAULT_POINTS_EACH[kind] || 0;
}

export function defaultGemValueEach(kind) {
  return kind === "gems" ? 1 : 0;
}

export function sanitizePaidItem(raw) {
  const kind = PAID_ITEM_KINDS.includes(String(raw?.kind)) ? String(raw.kind) : "custom";
  return {
    kind,
    name: String(raw?.name || "").trim().slice(0, 200),
    quantity: Math.max(0, Math.trunc(finite(raw?.quantity))),
    durationSeconds: paidKindHasTime(kind)
      ? Math.max(0, Math.trunc(finite(raw?.durationSeconds ?? raw?.duration_seconds)))
      : 0,
    gemValueEach: nonNegative(raw?.gemValueEach ?? raw?.gem_value_each),
    pointsEach: nonNegative(raw?.pointsEach ?? raw?.points_each, defaultPointsEach(kind)),
  };
}

export function sanitizePaidOffer(raw) {
  const base = emptyPaidOffer();
  return {
    offerId: String(raw?.offerId || raw?.offer_id || "").trim().slice(0, 100),
    title: String(raw?.title || "").trim().slice(0, 200),
    goal: PAID_GOALS.includes(String(raw?.goal)) ? String(raw.goal) : "all_round",
    memo: String(raw?.memo || "").slice(0, 2000),
    diamondCost: Math.max(0, Math.trunc(finite(raw?.diamondCost ?? raw?.diamond_cost))),
    includedGems: Math.max(0, Math.trunc(finite(raw?.includedGems ?? raw?.included_gems))),
    bonusGems: Math.max(0, Math.trunc(finite(raw?.bonusGems ?? raw?.bonus_gems))),
    items: (Array.isArray(raw?.items) ? raw.items : []).map(sanitizePaidItem),
    createdAt: String(raw?.createdAt || raw?.created_at || base.createdAt),
    updatedAt: String(raw?.updatedAt || raw?.updated_at || base.updatedAt),
  };
}

export function sanitizePaidValuation(raw) {
  const base = defaultPaidValuation();
  return {
    pointsPerGem: nonNegative(raw?.pointsPerGem ?? raw?.points_per_gem, base.pointsPerGem),
    generalSpeedupPointsPerHour: nonNegative(raw?.generalSpeedupPointsPerHour ?? raw?.general_speedup_points_per_hour, base.generalSpeedupPointsPerHour),
    researchSpeedupPointsPerHour: nonNegative(raw?.researchSpeedupPointsPerHour ?? raw?.research_speedup_points_per_hour, base.researchSpeedupPointsPerHour),
    trainingSpeedupPointsPerHour: nonNegative(raw?.trainingSpeedupPointsPerHour ?? raw?.training_speedup_points_per_hour, base.trainingSpeedupPointsPerHour),
    constructionSpeedupPointsPerHour: nonNegative(raw?.constructionSpeedupPointsPerHour ?? raw?.construction_speedup_points_per_hour, base.constructionSpeedupPointsPerHour),
    healingSpeedupPointsPerHour: nonNegative(raw?.healingSpeedupPointsPerHour ?? raw?.healing_speedup_points_per_hour, base.healingSpeedupPointsPerHour),
    mergingSpeedupPointsPerHour: nonNegative(raw?.mergingSpeedupPointsPerHour ?? raw?.merging_speedup_points_per_hour, base.mergingSpeedupPointsPerHour),
    craftingSpeedupPointsPerHour: nonNegative(raw?.craftingSpeedupPointsPerHour ?? raw?.crafting_speedup_points_per_hour, base.craftingSpeedupPointsPerHour),
    useSpeedupGemPresets: (raw?.useSpeedupGemPresets ?? raw?.use_speedup_gem_presets) !== false,
  };
}

export function summarizePaidOffer(rawOffer, rawValuation) {
  const offer = sanitizePaidOffer(rawOffer);
  const valuation = sanitizePaidValuation(rawValuation);
  const speedupSeconds = Object.fromEntries(SPEEDUP_ITEM_KINDS.map((kind) => [kind, 0]));
  let itemGemValue = 0;
  let speedupGemValue = 0;
  let directItemPoints = 0;
  for (const item of offer.items) {
    if (paidKindHasTime(item.kind)) {
      speedupSeconds[item.kind] += item.durationSeconds * item.quantity;
      if (valuation.useSpeedupGemPresets && item.gemValueEach <= 0) {
        const preset = item.kind === "merging"
          ? MERGING_SPEEDUP_GEM_VALUE_BY_SECONDS
          : item.kind === "crafting" ? null : SPEEDUP_GEM_VALUE_BY_SECONDS;
        speedupGemValue += (preset?.get(item.durationSeconds) || 0) * item.quantity;
      }
    }
    itemGemValue += item.gemValueEach * item.quantity;
    directItemPoints += item.pointsEach * item.quantity;
  }
  const includedGems = offer.includedGems + offer.bonusGems;
  const totalGemValue = includedGems + itemGemValue + speedupGemValue;
  const rates = {
    general: valuation.generalSpeedupPointsPerHour,
    research: valuation.researchSpeedupPointsPerHour,
    training: valuation.trainingSpeedupPointsPerHour,
    construction: valuation.constructionSpeedupPointsPerHour,
    healing: valuation.healingSpeedupPointsPerHour,
    merging: valuation.mergingSpeedupPointsPerHour,
    crafting: valuation.craftingSpeedupPointsPerHour,
  };
  const speedupPoints = SPEEDUP_ITEM_KINDS.reduce((sum, kind) => sum + speedupSeconds[kind] / 3600 * rates[kind], 0);
  const totalPoints = totalGemValue * valuation.pointsPerGem + directItemPoints + speedupPoints;
  return {
    speedupSeconds,
    totalSpeedupSeconds: Object.values(speedupSeconds).reduce((sum, value) => sum + value, 0),
    includedGems,
    itemGemValue,
    speedupGemValue,
    totalGemValue,
    directItemPoints,
    speedupPoints,
    totalPoints,
    pointsPerDiamond: offer.diamondCost > 0 ? totalPoints / offer.diamondCost : null,
    gemsPerDiamond: offer.diamondCost > 0 ? totalGemValue / offer.diamondCost : null,
  };
}

export function sortedPaidOffers(offers, valuation) {
  return [...(offers || [])].map(sanitizePaidOffer).sort((left, right) => {
    const a = summarizePaidOffer(left, valuation);
    const b = summarizePaidOffer(right, valuation);
    return (b.pointsPerDiamond || 0) - (a.pointsPerDiamond || 0)
      || b.totalPoints - a.totalPoints
      || left.title.localeCompare(right.title);
  });
}

export function paidOfferExchangePayload(offers, valuation, name = "") {
  return {
    document_type: PAID_OFFER_DOCUMENT_TYPE,
    schema_version: 1,
    exported_at: new Date().toISOString(),
    name: String(name || "").trim().slice(0, 200),
    valuation: {
      points_per_gem: valuation.pointsPerGem,
      general_speedup_points_per_hour: valuation.generalSpeedupPointsPerHour,
      research_speedup_points_per_hour: valuation.researchSpeedupPointsPerHour,
      training_speedup_points_per_hour: valuation.trainingSpeedupPointsPerHour,
      construction_speedup_points_per_hour: valuation.constructionSpeedupPointsPerHour,
      healing_speedup_points_per_hour: valuation.healingSpeedupPointsPerHour,
      merging_speedup_points_per_hour: valuation.mergingSpeedupPointsPerHour,
      crafting_speedup_points_per_hour: valuation.craftingSpeedupPointsPerHour,
      use_speedup_gem_presets: valuation.useSpeedupGemPresets,
    },
    offers: (offers || []).map((raw) => {
      const offer = sanitizePaidOffer(raw);
      return {
        offer_id: offer.offerId, title: offer.title, goal: offer.goal, memo: offer.memo,
        diamond_cost: offer.diamondCost, included_gems: offer.includedGems, bonus_gems: offer.bonusGems,
        items: offer.items.map((item) => ({ kind: item.kind, name: item.name, quantity: item.quantity, duration_seconds: item.durationSeconds, gem_value_each: item.gemValueEach, points_each: item.pointsEach })),
        created_at: offer.createdAt, updated_at: offer.updatedAt,
      };
    }),
  };
}

export function paidOffersFromExchangePayload(raw) {
  if (raw?.document_type !== PAID_OFFER_DOCUMENT_TYPE || Number(raw?.schema_version) !== 1 || !Array.isArray(raw?.offers)) throw new Error("対応していない課金データです");
  const offers = raw.offers.map(sanitizePaidOffer).filter((offer) => offer.title);
  if (!offers.length && (!raw.valuation || typeof raw.valuation !== "object" || Array.isArray(raw.valuation))) throw new Error("取り込める課金項目または比較設定がありません");
  return { offers, valuation: sanitizePaidValuation(raw.valuation), name: String(raw.name || "").slice(0, 200) };
}
